cli/config/credentials/oauth_store.go

```diff
@@ -0,0 +1,206 @@
+// returned as is. If the credentials retrieved do not parse as a token, they are
+// also returned as is.
+func (o *oauthStore) Get(serverAddress string) (types.AuthConfig, error) {
+	if serverAddress != defaultRegistry {
+		return o.backingStore.Get(serverAddress)
+	}
+
+	tokenRes := o.fetchFromBackingStore()
+	if tokenRes == nil {
+		return o.backingStore.Get(serverAddress)
+	}
+
+	// if the access token is valid for less than minimumTokenLifetime, refresh it
+	if tokenRes.RefreshToken != "" && tokenRes.Claims.Expiry.Time().Before(time.Now().Add(minimumTokenLifetime)) {
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		refreshRes, err := o.manager.RefreshToken(ctx, tokenRes.RefreshToken)
+		if err != nil || refreshRes == nil {
+			return types.AuthConfig{}, fmt.Errorf("failed to refresh token: %w", err)
+		}
+		tokenRes = refreshRes
+	}
+
+	err := o.storeInBackingStore(*tokenRes)
+	if err != nil {
+		return types.AuthConfig{}, err
+	}
+
+	return types.AuthConfig{
+		Username:      tokenRes.Claims.Domain.Username,
+		Password:      tokenRes.AccessToken,
+		ServerAddress: defaultRegistry,
+	}, nil
+}
+
+// GetAll returns a map of all credentials in the backing store. If the backing
+// store contains credentials for the official registry, these are refreshed/processed
+// according to the same rules as Get.
+func (o *oauthStore) GetAll() (map[string]types.AuthConfig, error) {
+	// fetch all authconfigs from backing store
+	allAuths, err := o.backingStore.GetAll()
+	if err != nil {
+		return nil, err
+	}
+
+	// if there are no oauth-type credentials for the default registry,
+	// we can return as-is
+	tokenRes := o.fetchFromBackingStore()
+	if tokenRes == nil {
+		return allAuths, nil
+	}
+
+	// if there is an oauth-type entry, then we need to parse it/refresh it
+	auth, err := o.Get(defaultRegistry)
+	if err != nil {
+		return nil, err
+	}
+	allAuths[defaultRegistry] = auth
+
+	// delete access/refresh-token specific entries since the caller
+	// doesn't care about those
+	delete(allAuths, accessTokenServerAddress)
+	delete(allAuths, refreshTokenServerAddress)
+
+	return allAuths, err
+}
+
+// Erase removes the credentials from the backing store.
+// If the address pertains to the default registry, and there are oauth-type
+// credentials stored, it also revokes the refresh token with the tenant.
+func (o *oauthStore) Erase(serverAddress string) error {
+	if serverAddress != defaultRegistry {
+		return o.backingStore.Erase(serverAddress)
+	}
+
+	refreshTokenAuth, err := o.backingStore.Get(refreshTokenServerAddress)
+	if err == nil && refreshTokenAuth.Password != "" {
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		err = o.manager.Logout(ctx, refreshTokenAuth.Password)
+		if err != nil {
+			// todo(laurazard): actual message here
+			fmt.Fprint(os.Stderr, "Failed to revoke refresh token with tenant.. Credentials will still be erased.")
```

Comment:
```suggestion
			_, _ = fmt.Fprint(os.Stderr, "Failed to revoke refresh token with tenant.. Credentials will still be erased.")
```
(to be consistent with other Fprints)

Reply:
What about making this a warning? I don't recall if we have a helper function for styling, but something like:

```suggestion
			_, _ = fmt.Fprint(os.Stderr, "Warning: You have been logged out successfully, but there was a failure to revoke the OAuth refresh token with the tenant.")
```

The only problem with the above is that it doesn't really say anything about manually addressing that failure. A more complete, but unattractively longer, version could be:

```suggestion
			_, _ = fmt.Fprint(os.Stderr, "Warning: You have been logged out successfully, but there was a failure to revoke the OAuth refresh token with the tenant. You may want to manually revoke the token through your OAuth provider's settings.")
```

Reply:
I don't think the `You have been logged out successfully` works super well with the messages we have [here](https://github.com/docker/cli/blob/0bac5a707b23b588a379ca1a05a0bf5a77155c32/cli/command/registry/logout.go#L66) and [here](https://github.com/docker/cli/blob/0bac5a707b23b588a379ca1a05a0bf5a77155c32/cli/command/registry/logout.go#L56).

In those, we don't say "logging out", but rather `could not erase credentials` or `Removing login credentials`. @dvdksn

Reply:
Wasn't aware of the prior art 👍🏻 In that case `s/You have been logged out successfully/Credentials erased successfully/` for consistency.